src/background/modules/CssAnalysis.js

```diff
@@ -3,77 +3,47 @@ import * as BrowserCommunication from "/common/modules/BrowserCommunication/Brow
 import { COMMUNICATION_MESSAGE_TYPE } from "/common/modules/data/BrowserCommunicationTypes.js";
 
 const TAB_FILTER_URLS = ["<all_urls>"];
-let lastSettingsInjection = null;
+let lastContentScript = null;
 
 AddonSettings.setCaching(false);
 
-/**
- * Triggers the new setting in case it has been changed.
- *
- * @public
- * @returns {Promise}
- */
-export async function triggerNewColorStatus() {
-    // add new code injection and remove old one
-    if (lastSettingsInjection) {
-        (await lastSettingsInjection).unregister();
-    }
-    lastSettingsInjection = await enableSettingInjection();
-}
-
 /**
  * Adds a content script that is injected by Firefox and provides the preloaded
  * value of fakedColorStatus.
  *
  * @private
+ * @param {string} [fakedColorStatus]
  * @returns {Promise}
  */
-async function enableSettingInjection() {
-    const functionalMode = await AddonSettings.get("functionalMode");
-    const fakedColorStatus = await AddonSettings.get("fakedColorStatus");
+export async function injectContentScript(fakedColorStatus) {
+    const settings = {
+        fakedColorStatus: fakedColorStatus || await AddonSettings.get("fakedColorStatus"),
+        functionalMode: await AddonSettings.get("functionalMode")
+    };
 
-    return browser.contentScripts.register({
+    // add new code injection
+    const newContentScript = await browser.contentScripts.register({
         matches: TAB_FILTER_URLS,
         js: [{
-            code: `
-                    ${setSettings.toString()}
-                    setSettings(${functionalMode}, "${fakedColorStatus.toUpperCase()}");
+            code: `;
+                    var initialSettings = JSON.parse('${JSON.stringify(settings)}');
```

Comment:
> serializing all options

The "settings" there is not all settings, just the ones needed for correct operation at document-start, which, as things currently stands, both `fakedColorStatus` and `functionalMode` are.
Technically, `functionalMode` is not strictly required, but until "applyColorSchemeCss.js" is improved, specifically `applyWantedStyle()`, you will have a race condition there, because `DOMContentLoaded` might be fired faster than any `browser.storage.sync.get` promises resolve.
So, I'm serializing both of them here.

In the future, if you have settings that does not need to be available at document-start, just don't add them to `const settings = { ... }` above that.
I think `injectContentScript()` is a relatively infrequent operation (one is not likely to rapidly change settings), so `JSON.stringify` is only infrequently called; and the performance of `JSON.parse` vs object initializer parsing is negligable for one off things like this.

I don't think the performance angle is significant here, and `JSON.parse('{...}')` is *much* safer than `({...})`.

Reply:
> I think injectContentScript() is a relatively infrequent operation (one is not likely to rapidly change settings), so JSON.stringify is only infrequently called

Fair enough, but let me think…
IIRC it _is_ called when a new tab is opened.
As such, it _could_ be a problem if you open a website in a new tab.

That said, if you've tested it and you really see no performance impact by that (I have no idea how these JS features really are), then okay. I'll totally believe you then. :wink:

Reply:
> IIRC it is called when a new tab is opened.

`injectContentScript` is called from 4 places:
- `export async function init` in CssAnalysis.js, which is at extension start.
- The two Listeners for COMMUNICATION_MESSAGE_TYPE.NEW_SETTING and COMMUNICATION_MESSAGE_TYPE.NEW_ADDIONAL_SETTINGS.
- `function propagateNewSetting` in ActionButton.js, that is itself only called from `browser.browserAction.onClicked`.

Nothing on tab creation.

And, it is `JSON.parse` that is called whenever a window is created (e.g. opening a web page or a tab), not `JSON.stringify`.
See:
```
function testInner() {
 console.log('called');
 return 'bar';
}
function testCallback() {
 console.log(window.testtest);
}
function testOuter() {
 window.testtest = `before ${testInner()} after`;
 setTimeout(testCallback, 1000);
}
testOuter();
```